<a href="https://colab.research.google.com/github/jubayer1608015/labtask/blob/main/1608015_labtask02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Experiment Name: Exploring Different Types of Regression

Course No: ETE 456

Name: Khondokar Jubayer Rahman
ID: 1608015

Objective

1.To apply different regression algorithms in real world dataset.

Task

Prediction of Insurance Costs

In [44]:
import warnings
warnings.filterwarnings("ignore")


Importing The Libraries

In [45]:
import numpy as np        
import pandas as pd     
import matplotlib.pyplot as plt


In [46]:
# Download the data
!wget -O insurance.csv https://www.dropbox.com/s/mwgqgjbmfw0xa5p/insurance.csv?dl=0

--2021-12-21 08:07:49--  https://www.dropbox.com/s/mwgqgjbmfw0xa5p/insurance.csv?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.2.18, 2620:100:6017:18::a27d:212
Connecting to www.dropbox.com (www.dropbox.com)|162.125.2.18|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/mwgqgjbmfw0xa5p/insurance.csv [following]
--2021-12-21 08:07:49--  https://www.dropbox.com/s/raw/mwgqgjbmfw0xa5p/insurance.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc75618bc07f2636a1b777c19390.dl.dropboxusercontent.com/cd/0/inline/BcSj3X_PYUeWAmgcpg2zn4c0KJHqXO1otohQcDxum82l7yV2G71ntBhbOrRc764UHaHDb2jD7LIfYiXlXNGqYoXxzRIAAPBJ-EIZQx5_L1wCIwlthFrs9ypzlynQundxX72tSMFMNBWuHp2_26IMPsBV/file# [following]
--2021-12-21 08:07:49--  https://uc75618bc07f2636a1b777c19390.dl.dropboxusercontent.com/cd/0/inline/BcSj3X_PYUeWAmgcpg2zn4c0KJHqXO1otohQcDxum82l7yV2G71ntBhbOrRc764UHaHDb2jD7LIfYiXlXN

In [47]:
dataset = pd.read_csv('insurance.csv')
dataset

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

Now splitting the dataset into independent and dependent variables. Independent are the features which are used for prediction. Dependent is the target which will be predicted.

In [48]:
features = dataset[['age'	,'sex',	'bmi','children','smoker','region']]  # for independent
target = dataset[['charges']]   # for dependent

In [49]:
features

age     sex     bmi  children smoker     region
0      19  female  27.900         0    yes  southwest
1      18    male  33.770         1     no  southeast
2      28    male  33.000         3     no  southeast
3      33    male  22.705         0     no  northwest
4      32    male  28.880         0     no  northwest
...   ...     ...     ...       ...    ...        ...
1333   50    male  30.970         3     no  northwest
1334   18  female  31.920         0     no  northeast
1335   18  female  36.850         0     no  southeast
1336   21  female  25.800         0     no  southwest
1337   61  female  29.070         0    yes  northwest

[1338 rows x 6 columns]

In [50]:
target

charges
0     16884.92400
1      1725.55230
2      4449.46200
3     21984.47061
4      3866.85520
...           ...
1333  10600.54830
1334   2205.98080
1335   1629.83350
1336   2007.94500
1337  29141.36030

[1338 rows x 1 columns]

Taking Care of Missing Data

Here 'sklearn' means sckit learn which is a package used for machine learning model. It contains many libraries.

preprocessing is a library which contains several classes,methods. From this library we import the Imputer class which will allow us to take care of the missing data

In [51]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values=np.nan,strategy = "mean") # imputer is an object of Imputer class 
imputer = imputer.fit(features[['age','bmi','children']])

In [52]:
features[['age','bmi','children']]= imputer.transform(features[['age','bmi','children']])

In [53]:
features

age     sex     bmi  children smoker     region
0     19.0  female  27.900       0.0    yes  southwest
1     18.0    male  33.770       1.0     no  southeast
2     28.0    male  33.000       3.0     no  southeast
3     33.0    male  22.705       0.0     no  northwest
4     32.0    male  28.880       0.0     no  northwest
...    ...     ...     ...       ...    ...        ...
1333  50.0    male  30.970       3.0     no  northwest
1334  18.0  female  31.920       0.0     no  northeast
1335  18.0  female  36.850       0.0     no  southeast
1336  21.0  female  25.800       0.0     no  southwest
1337  61.0  female  29.070       0.0    yes  northwest

[1338 rows x 6 columns]

In [54]:
imputer = SimpleImputer(missing_values=np.nan,strategy = "mean") # imputer is an object of Imputer class 
imputer = imputer.fit(target[['charges']])

In [55]:
target[['charges']]= imputer.transform(target[['charges']])

In [56]:
target

charges
0     16884.92400
1      1725.55230
2      4449.46200
3     21984.47061
4      3866.85520
...           ...
1333  10600.54830
1334   2205.98080
1335   1629.83350
1336   2007.94500
1337  29141.36030

[1338 rows x 1 columns]

In [57]:
features['sex'].unique()

array(['female', 'male'], dtype=object)

In [58]:
features['sex'].value_counts()

male      676
female    662
Name: sex, dtype: int64

In [59]:
features['smoker'].unique()

array(['yes', 'no'], dtype=object)

In [60]:
features['smoker'].value_counts()

no     1064
yes     274
Name: smoker, dtype: int64

In [61]:
features['region'].unique()

array(['southwest', 'southeast', 'northwest', 'northeast'], dtype=object)

Encoding Categorical Data

One Hot Encoding

One hot encoder is used to convert the categorical data into numerical, so that they can be used in prediction.

In [62]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder

In [63]:
encoder=OneHotEncoder(sparse=False)
encoded_labels = pd.DataFrame (encoder.fit_transform(features[['sex','smoker','region']]))

In [64]:
encoded_labels.columns = encoder.get_feature_names(['sex','smoker','region'])
dataset= pd.concat([features, encoded_labels ], axis=1)

In [65]:
dataset

age     sex  ...  region_southeast  region_southwest
0     19.0  female  ...               0.0               1.0
1     18.0    male  ...               1.0               0.0
2     28.0    male  ...               1.0               0.0
3     33.0    male  ...               0.0               0.0
4     32.0    male  ...               0.0               0.0
...    ...     ...  ...               ...               ...
1333  50.0    male  ...               0.0               0.0
1334  18.0  female  ...               0.0               0.0
1335  18.0  female  ...               1.0               0.0
1336  21.0  female  ...               0.0               1.0
1337  61.0  female  ...               0.0               0.0

[1338 rows x 14 columns]

In [66]:
dataset.columns

Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'sex_female',
       'sex_male', 'smoker_no', 'smoker_yes', 'region_northeast',
       'region_northwest', 'region_southeast', 'region_southwest'],
      dtype='object')

In [67]:
new_features = dataset[['age', 'sex_female', 'sex_male',
       'smoker_no', 'smoker_yes', 'bmi','children','region_northeast','region_northwest','region_southeast','region_southwest']]

In [68]:
from sklearn.model_selection import train_test_split

In [69]:
X_train,X_test,y_train,y_test=train_test_split(new_features,target,test_size=0.1,random_state=73)
# random_state = 0 is select to get the same result

In [70]:
print(X_train.shape)
print(X_test.shape)

(1204, 11)
(134, 11)


In [71]:
X_train

age  sex_female  ...  region_southeast  region_southwest
204   46.0         0.0  ...               0.0               1.0
703   34.0         1.0  ...               0.0               0.0
24    37.0         0.0  ...               0.0               0.0
1222  50.0         0.0  ...               1.0               0.0
106   19.0         1.0  ...               0.0               1.0
...    ...         ...  ...               ...               ...
321   26.0         1.0  ...               0.0               0.0
323   57.0         0.0  ...               0.0               0.0
394   46.0         1.0  ...               0.0               0.0
146   46.0         0.0  ...               0.0               0.0
918   61.0         1.0  ...               0.0               1.0

[1204 rows x 11 columns]

Fitting Multiple Linear Regression Model to Training Set

Different types of Regression Algorithm

Linear Regression (Univariate or Multivariate)

 Support Vector Regression

  Decision Tree Regression 

  Random Forest Regressrion

In [72]:
from sklearn.linear_model import LinearRegression

regressor = LinearRegression()
regressor.fit(X_train,y_train)

# Predicting the Test Set Results
y_pred = regressor.predict(X_test)

In [73]:
y_test

charges
803   38792.68560
263   36219.40545
196    5649.71500
694    3577.99900
37     2302.30000
...           ...
986    8410.04685
258   11520.09985
49    38709.17600
1079  15161.53440
1228  10594.22570

[134 rows x 1 columns]

In [74]:
y_pred

array([[29669.37089159],
       [28823.93374863],
       [ 8166.09658616],
       [ 6168.11407507],
       [  799.60702066],
       [25670.79275768],
       [10536.29379753],
       [11198.2978864 ],
       [11390.55535189],
       [39110.75057103],
       [13060.42139099],
       [30971.7066379 ],
       [32032.91626116],
       [ 3730.67672065],
       [ 7855.42051837],
       [32119.08962045],
       [ 7231.36619798],
       [31931.03209881],
       [ 2585.94975287],
       [33812.93660703],
       [ 8146.96517646],
       [ 9369.733045  ],
       [  -51.95234889],
       [13905.72688798],
       [ 5713.93107343],
       [ 2394.45462181],
       [34332.16736543],
       [32411.63659129],
       [ 1047.51668472],
       [ 5649.38794692],
       [15681.52113764],
       [ 4369.01727831],
       [35205.59745212],
       [ 6873.63302937],
       [ 3710.93323686],
       [29040.04869295],
       [15218.81701735],
       [ 8843.2776472 ],
       [16744.76717616],
       [-1006.31185285],


Evaluation Metrices

Mean Absolute Error (MAE) Mean Squared Error (MSE) R-Squared Error

In [75]:
from sklearn.metrics import mean_absolute_error

# MAE

mean_absolute_error(y_test, y_pred)

4640.5878994470695

In [76]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_test, y_pred)

46853261.17967038

In [77]:
from sklearn.metrics import r2_score

r2_score(y_test, y_pred)

0.7113969820315541

In [78]:
dataset

age     sex  ...  region_southeast  region_southwest
0     19.0  female  ...               0.0               1.0
1     18.0    male  ...               1.0               0.0
2     28.0    male  ...               1.0               0.0
3     33.0    male  ...               0.0               0.0
4     32.0    male  ...               0.0               0.0
...    ...     ...  ...               ...               ...
1333  50.0    male  ...               0.0               0.0
1334  18.0  female  ...               0.0               0.0
1335  18.0  female  ...               1.0               0.0
1336  21.0  female  ...               0.0               1.0
1337  61.0  female  ...               0.0               0.0

[1338 rows x 14 columns]

Feature Scaling

It is necessary because sometimes a large value in the variable can dominate the small value.

In [79]:
from sklearn.preprocessing import StandardScaler
X_sc = StandardScaler()
y_sc = StandardScaler()
X_train_new = X_sc.fit_transform(X_train[['age', 'sex_female', 'sex_male',
       'smoker_no', 'smoker_yes', 'bmi','children','region_northeast','region_northwest','region_southeast','region_southwest']])
X_test_new = X_sc.transform(X_test[['age', 'sex_female', 'sex_male',
       'smoker_no', 'smoker_yes', 'bmi','children','region_northeast','region_northwest','region_southeast','region_southwest']])
y_train_new = y_sc.fit_transform(y_train[['charges']])
y_test_new = y_sc.transform(y_test[['charges']])

Support Vector Regression

In [80]:
from sklearn.svm import SVR 

regressor = SVR(kernel = 'linear')
regressor.fit(X_train_new,y_train_new)


SVR(kernel='linear')

In [81]:
y_pred = regressor.predict(X_test_new)

In [39]:
r2_score(y_test_new, y_pred)

0.7619821999423445

Decision Tree Regression

In [83]:
from sklearn.tree import DecisionTreeRegressor


regressor = DecisionTreeRegressor(random_state = 73)
regressor.fit(X_train_new,y_train_new)

DecisionTreeRegressor(random_state=73)

In [84]:
y_pred = regressor.predict(X_test_new)
r2_score(y_test_new, y_pred)

0.7599587544627731

Random Forest Regression

In [86]:
from sklearn.ensemble import RandomForestRegressor

regressor = RandomForestRegressor(n_estimators = 12,random_state = 73) # n estiamator is the number of decision trees
regressor.fit(X_train_new,y_train_new)

RandomForestRegressor(n_estimators=12, random_state=73)

In [87]:
y_pred = regressor.predict(X_test_new)
r2_score(y_test_new, y_pred)

0.834483315682588

The R-square error for the regression are given below:

Linear Regression :0.7113969820315541

Support Vector Regression: 0.7619821999423445

Decision Tree Regression:0.7599587544627731

Random Forest Regressrion: 0.834483315682588

In the dataset, Multivariate Linear Regression, Support Vector Regression, Decision Tree Regression, Random Forest Regressrion has been applied. For evaluating their performances R-square error evaluation matrix has been applied for all. From the values we can see that, R-square value for random forest regression is more close to 1 which is 0.85 than the others. So that, random forest regression gives better prediction in the dataset.